swarmrl/observables/subdivided_vision_cones.py

```diff
@@ -0,0 +1,165 @@
+
+    def __init__(
+        self, vision_range: float, vision_half_angle: float, n_cones: int, radii=[]
+    ):
+        """
+        Constructor for the observable.
+
+        Parameters
+        ----------
+        vision_range : float
+                How far the particles can see.
+        vision_half_angle : float
+                the half width of the field of view in radiant units
+        n_cones: int
+                In how many cone is the field of view subdivided
+        radii: list
+                List of the radii of the colloids in the experiment,
+                ordered by the ids of the colloids
+        """
+        self.vision_range = vision_range
+        self.vision_half_angle = vision_half_angle
+        self.n_cones = n_cones
+        self.radii = radii
+        self.types = []
+
+    def initialize(self, colloids: List[Colloid]):
+        """
+        This method is not needed for the vision_cone because the initialization
+        currently only initialises general features for all colloids while the vision
+        cone needs to be calculated for each cone separately.
+        """
+        pass
+
+    def detect_things_to_see(self, colloids: List[Colloid]):
+        """
+        Calculates the normalised director of the colloids.
+
+        Parameters
+        ----------
+        colloids : object
+                Colloids with all possible different types.
+
+        Returns
+        all possible types in the corresponding index they will also later be found in,
+        when the observable is calculated
+        """
+        types = []
+        for c in colloids:
+            if c.type not in types:
+                types.append(c.type)
+        self.types = np.array(types)
+
+    def _calculate_director(self, colloid):
+        """
+        Calculates the normalised director of the colloids.
+
+        Parameters
+        ----------
+        colloid : object
+                Colloid for which the observable should be computed.
+
+        Returns
+        Colloid position and vision cone director.
+        """
+        my_pos = np.copy(colloid.pos[:2])
+        my_director = np.copy(colloid.director[:2])
+        # TODO: Add 3D support
+        my_director = my_director / np.linalg.norm(my_director)
+        return my_pos, my_director
+
+    def _calculate_cones(self, my_pos, my_director, colloids: List[Colloid]):
+        """
+        Calculates the vision cones of the colloid.
+
+        Parameters
+        ----------
+        my_pos : np.ndarray
+                Position of the colloid.
+        my_director : np.ndarray
+                Normalised director of the colloid.
+        colloids : List[Colloid]
+                all colloids
+        Returns
+        np.array of shape (n_cones, num_of_types) containing the vision values
+        for each cone and for each particle type that can be visible.
+        """
+        vision_val_out = np.zeros((self.n_cones, len(self.types)))
+
+        for c in colloids:
```

Comment:
We could probably parallelised this using `jax.vmap` but perhaps this can be looked at later as it is not critical at the moment.